scipy/special/_basic.py

```diff
@@ -3182,63 +3229,10 @@ def factorial2(n, exact=False, extend="zero"):
     .. [1] Complex extension to double factorial
             https://en.wikipedia.org/wiki/Double_factorial#Complex_arguments
     """
-    if extend not in ("zero", "complex"):
-        raise ValueError(
-            f"argument `extend` must be either 'zero' or 'complex', received: {extend}"
-        )
-    if exact and extend == "complex":
-        raise ValueError("Incompatible options: `exact=True` and `extend='complex'`")
+    return _factorialx_wrapper("factorial2", n, k=2, exact=exact, extend=extend)
 
-    msg_needs_complex = (
-        "In order to use non-integer arguments, you must opt into this by passing "
-        "`extend='complex'`. Note that this changes the result for all negative "
-        "arguments (which by default return 0). Additionally, it will rescale the "
-        "values of the double factorial at even integers by a factor of sqrt(2/pi)."
-    )
-    msg_wrong_dtype = (
-        "Unsupported data type for factorial2: {dtype}\n"
-        "Only integers are permitted by default, though floating point "
-        "and complex numbers can be used if `extend='complex'` is passed."
-    )
 
-    # don't use isscalar due to numpy/numpy#23574; 0-dim arrays treated below
-    if np.ndim(n) == 0 and not isinstance(n, np.ndarray):
-        # scalar cases
-        if not _is_subdtype(type(n), ["i", "f", "c", type(None)]):
-            raise ValueError(msg_wrong_dtype.format(dtype=type(n)))
-        elif _is_subdtype(type(n), ["f", "c"]) and extend != "complex":
-            raise ValueError(msg_needs_complex)
-        elif n is None or np.isnan(n):
-            complexify = (extend == "complex") and _is_subdtype(type(n), "c")
-            return np.complex128("nan+nanj") if complexify else np.float64("nan")
-        elif extend == "zero" and n < 0:
-            return 0 if exact else np.float64(0)
-        elif n in {0, 1}:
-            return 1 if exact else np.float64(1)
-
-        if exact:
-            # general integer case
-            return _range_prod(1, n, k=2)
-        # approximation
-        return _factorialx_approx_core(n, k=2, extend=extend)
-
-    # arrays & array-likes
-    n = asarray(n)
-
-    if not _is_subdtype(n.dtype, ["i", "f", "c"]):
-        raise ValueError(msg_wrong_dtype.format(dtype=n.dtype))
-    elif _is_subdtype(n.dtype, ["f", "c"]) and extend != "complex":
-        raise ValueError(msg_needs_complex)
-
-    if n.size == 0:
-        # return empty arrays unchanged
-        return n
-    elif exact:
-        return _factorialx_array_exact(n, k=2)
-    return _factorialx_array_approx(n, k=2, extend=extend)
-
-
-def factorialk(n, k, exact=None, extend="zero"):
+def factorialk(n, k, exact=False, extend="zero"):
```

Comment:
Here and elsewhere should we make the new input keyword only?

```suggestion
def factorialk(n, k, exact=False, *, extend="zero"):
```

Reply:
I thought about it - not sure. If we do it should IMO be at the [right](https://github.com/scipy/scipy/issues/18703#issuecomment-1659814261) place, not just for the new keyword. I even looked up my mailing list [post](https://mail.python.org/archives/list/scipy-dev@python.org/thread/MK7AWN5ARX7AOJXDANUPTFYHSJ5Y53ES/) on this (already >1y 😱). I guess we'd be ready for a large-scale change-over, as soon as people are comfortable with discussing SciPy 2.0...

Reply:
Fair point